Fix most violations of new CA1859

stephentoub opened the pull request:
The analyzer recommends replacing uses of types with other types that could reduce overheads (e.g. `List<T>` instead of `IList<T>`).  This fixes most of the violations we currently have of the rule, but it doesn't enable it yet because there are too many false positives; fixes for those are in-flight.  Some of these replacements won't help with perf (e.g. using `ArgumentException` instead of `Exception` as the return type of a throw helper create method), but there's little harm to them, and some of them do avoid overheads like allocation (e.g. foreach'ing something typed as `Dictionary<>` instead of `IDictionary<>` will avoid an enumerator allocation, accessing `Count` on a `List<T>` instead of `IList<T>` will avoid an interface dispatch, etc.)

cc: @buyaa-n, @geeknoid

geeknoid commented:
Wow, that's a lot of stuff.

CA1859 currently only recommends changing the signature of private methods and fields, the idea being that the changes are completely local to a given type. Do you think it would be useful to extend the analysis to internal symbols too, at least optionally? I bet that would raise another crop of these recommendations.

Another thing I'd like to add in the analyzer is a basic understanding of collections. So if you have an List<IFoo> but you're always assigning Foo instances to it, it would recommend upgrading to a List<Foo>. I think that would be another frequent hit.

stephentoub committed:
Fix most violations of new CA1859

The analyzer recommends replacing uses of types with other types that could reduce overheads (e.g. `List<T>` instead of `IList<T>`).  This fixes most of the violations we currently have of the rule, but it doesn't enable it yet because there are too many false positives; fixes for those are in-flight.  Some of these replacements won't help with perf (e.g. using `ArgumentException` instead of `Exception` as the return type of a throw helper create method), but there's little harm to them, and some of them do avoid overheads like allocation (e.g. foreach'ing something typed as `Dictionary<>` instead of `IDictionary<>` will avoid an enumerator allocation, accessing `Count` on a `List<T>` instead of `IList<T>` will avoid an interface dispatch, etc.)

stephentoub commented:
> CA1859 currently only recommends changing the signature of private methods and fields, the idea being that the changes are completely local to a given type. Do you think it would be useful to extend the analysis to internal symbols too, at least optionally?

Can you share an example?

> Another thing I'd like to add in the analyzer is a basic understanding of collections.

Yeah, I think that would be useful, at least for some of the most common, like IEnumerable and List.

buyaa-n commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509Pal.OpenSsl.cs:
```diff
@@ -5,9 +5,6 @@ namespace System.Security.Cryptography.X509Certificates
-        {
-            return new OpenSslX509Encoder();
-        }
+        private static partial IX509Pal BuildSingleton() => new OpenSslX509Encoder();
```
NIT
```suggestion
        private static partial OpenSslX509Encoder BuildSingleton() => new OpenSslX509Encoder();
```

stephentoub replied:
The return type here can't be changed as it's a partial method and the signature of all of the partials need to match, but the partial declaration is in the cross-platform shared file.

buyaa-n reviewed (approved):
Left a few NITs, overall LGTM, thank you!

stephentoub commented on src/libraries/System.Net.HttpListener/src/System/Net/Windows/HttpRequestStream.Windows.cs:
```diff
@@ -199,9 +199,8 @@ private void UpdateAfterRead(uint statusCode, uint dataRead)
-                        Exception exception = new HttpListenerException((int)statusCode);
+                        var exception = new HttpListenerException((int)statusCode);
                         if (NetEventSource.Log.IsEnabled()) NetEventSource.Error(this, exception.ToString());
-                        asyncResult.InternalCleanup();
```
> was it removed intentionally?

Yes, `asyncResult.InternalCleanup()` was just called on L194.

stephentoub committed:
Merge branch 'main' into ca1859

stephentoub committed:
Fix mistaken replacement

stephentoub commented on src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/Error.cs:
```diff
@@ -14,1057 +14,1057 @@ internal static class Error
+        internal static ArgumentException NoOrInvalidRuleProduced()
         {
-            return new InvalidOperationException(Strings.NoOrInvalidRuleProduced);
+            return new ArgumentException(Strings.NoOrInvalidRuleProduced);
```
> was this change intentional?

No, this was a mistake. Fixed. Thanks.

geeknoid commented:
Be gone Peanut Butter!
